### PR TITLE
Make PS compatible with PHP 7.2 

### PR DESCRIPTION
**Pull request by Quetzacoalt91**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | When testing PrestaShop with PHP 7.2, many errors appeared in the PHP error log. This PR begins the work to do with this version.
| Type?         | new feature
| Category?     | CO
| BC breaks?    | Nope
| Deprecations? | Nope
| Fixed ticket? | http://forge.prestashop.com/browse/BOOM-4444
| How to test?  | Test on a server running with PHP 7.2.

Related PRs:
- [x] https://github.com/PrestaShop/ps_facetedsearch/pull/30
- [x] https://github.com/PrestaShop/ps_linklist/pull/47

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/8246)

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -21,6 +21,7 @@ php:
   - 5.4
   - 5.6
   - 7.0
+  - nightly
```
Comment (kokspflanze) [inline .travis.yml]: nightly is atm php 7.3, you have to add 7.2 for that php version

Comment (Quetzacoalt91) [reply .travis.yml]: When we look at the documentation, 7.2 seem to be unavailable. https://docs.travis-ci.com/user/languages/php#Choosing-PHP-versions-to-test-against

I give it a try anyway, I will bring back `nightly` if it doesn't work.

Comment (kokspflanze) [reply .travis.yml]: yes doc looks bit old, but if you check https://travis-ci.org/doctrine/DoctrineORMModule and https://github.com/doctrine/DoctrineORMModule/blob/master/.travis.yml#L48
there you can see that php 7.2 already added,

**Comment by eternoendless** [comment]
Reviewed 26 of 30 files at r1, 13 of 13 files at r2.
Review status: 35 of 39 files reviewed at latest revision, 3 unresolved discussions.

---

*[classes/controller/AdminController.php, line 566 at r2](https://reviewable.io:443/reviews/prestashop/prestashop/8246#-L2QXxBV6rlK3B62wUsf:-L2QXxBV6rlK3B62wUsg:b-jv5d8t) ([raw file](https://github.com/prestashop/prestashop/blob/57128bb77002fd0839dc8fb2b37eecf415c59319/classes/controller/AdminController.php#L566)):*
> ```PHP
>      * @param int|null $tab_id
>      */
>     public function initBreadcrumbs($tab_id = null)
> ```

We shouldn't change the signature of public methods until the next major version

---

*[classes/controller/AdminController.php, line 572 at r2](https://reviewable.io:443/reviews/prestashop/prestashop/8246#-L2QYBXDu2WlfO-D1xDm:-L2QYBXDu2WlfO-D1xDn:b-28qa15) ([raw file](https://github.com/prestashop/prestashop/blob/57128bb77002fd0839dc8fb2b37eecf415c59319/classes/controller/AdminController.php#L572)):*
> ```PHP
>         }
> 
>         $tabs = Tab::recursiveTab($tab_id, array());
> ```

Why remove the $tab parameter?

---

*[classes/controller/AdminController.php, line 3512 at r2](https://reviewable.io:443/reviews/prestashop/prestashop/8246#-L2QYTpkXTvG5XsKsHbt:-L2QYTpkXTvG5XsKsHbu:b-algn8x) ([raw file](https://github.com/prestashop/prestashop/blob/57128bb77002fd0839dc8fb2b37eecf415c59319/classes/controller/AdminController.php#L3512)):*
> ```PHP
>         }
> 
>         if (is_null($filter_modules_list) || !count($filter_modules_list)) {
> ```

Shouldn't this be `!is_array($filter_modules_list) || !count(...)`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8246)*

**Comment by Quetzacoalt91** [comment]
*[classes/controller/AdminController.php, line 3512 at r2](https://reviewable.io:443/reviews/prestashop/prestashop/8246#-L2QYTpkXTvG5XsKsHbt:-L2dgD9AhlMaGgwNKHuZ:b-e72v21) ([raw file](https://github.com/prestashop/prestashop/blob/57128bb77002fd0839dc8fb2b37eecf415c59319/classes/controller/AdminController.php#L3512)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>

Shouldn't this be `!is_array($filter_modules_list) || !count(...)`?
</blockquote></details>

`$filter_modules_list` can be an array or null. That's an opportunity to remind this (tests were broken without that change).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8246)*

**Comment by Quetzacoalt91** [comment]
Review status: 34 of 39 files reviewed at latest revision, 3 unresolved discussions.

---

*[classes/controller/AdminController.php, line 566 at r2](https://reviewable.io:443/reviews/prestashop/prestashop/8246#-L2QXxBV6rlK3B62wUsf:-L2dg6yvPGdWn4hs7dcF:b-3fnqf2) ([raw file](https://github.com/prestashop/prestashop/blob/57128bb77002fd0839dc8fb2b37eecf415c59319/classes/controller/AdminController.php#L566)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>

We shouldn't change the signature of public methods until the next major version
</blockquote></details>

You're totally right. Reverted

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8246)*

**Comment by Quetzacoalt91** [comment]
Comments taken in account. Do not forget to check the related module PR, needed to make this one working on Travis.

**Comment by LittleBigDev** [comment]
:lgtm:

---

Reviewed 4 of 4 files at r3, 1 of 1 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8246#-:-L5K82Zo0FiPshlzXCHv:bnfp4nl)*

**Review by LittleBigDev — DISMISSED** [review]
Thank you @Quetzacoalt91

**Comment by eternoendless** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[classes/controller/AdminController.php, line 3512 at r2](https://reviewable.io:443/reviews/prestashop/prestashop/8246#-L2QYTpkXTvG5XsKsHbt:-L5NwjfVpDf0XJMPhpYx:b-w9sxr8) ([raw file](https://github.com/prestashop/prestashop/blob/57128bb77002fd0839dc8fb2b37eecf415c59319/classes/controller/AdminController.php#L3512)):*
<details><summary><i>Previously, Quetzacoalt91 (Thomas N) wrote…</i></summary><blockquote>

`$filter_modules_list` can be an array or null. That's an opportunity to remind this (tests were broken without that change).
</blockquote></details>

Ok then

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8246#-:-L5NwvdDqNX7_Tg7wE0q:bnfp4nl)*

**Comment by eternoendless** [comment]
After a closer look, it looks like `AdminControllerCore::setMedia()` has a different signature than `ControllerCore::setMedia()`, that will probably need fixing too

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8246#-:-L5OZ0jhjCQ6QLWHsocu:b-ok9boi)*

**Comment by Quetzacoalt91** [comment]
@eternoendless --> See above https://github.com/PrestaShop/PrestaShop/pull/8246#issuecomment-357187489

**Comment by Quetzacoalt91** [comment]
It seems not having the same signature as an abstract class is not an issue, fortunately. Otherwise we'd have to update all the front controllers, although they do not need this variable.

Tested with

```php
abstract class Controller
{
    abstract public function setMedia();
}

class AdminController extends Controller
{
    public function setMedia($isNewTheme = false)
    {
        echo "AdminController\n";
    }
}

class FrontController extends Controller
{
    public function setMedia()
    {
        echo "FrontController\n";
    }
}

class MyController extends AdminController
{
    public function setMedia($isNewTheme = false)
    {
        echo "MyController\n";
    }
}

(new AdminController)->setMedia();
(new FrontController)->setMedia();
(new MyController)->setMedia();
```

**Comment by LittleBigDev** [comment]
:lgtm:

---

Reviewed 36 of 37 files at r6, 1 of 2 files at r7, 1 of 1 files at r8.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8246#-:-L5y9SIkwNfgSsd5VXKe:bnfp4nl)*

**Review by LittleBigDev — DISMISSED** [review]

**Comment by Quetzacoalt91** [comment]
Close in favor of #8799

**Review by mickaelandrieu — APPROVED** [review]

**Review by mickaelandrieu — DISMISSED** [review]
This pull request needs to be finished before 1.7.4.0 release

**Comment by eternoendless** [comment]
:lgtm:

---

Reviewed 36 of 37 files at r6, 2 of 2 files at r7, 12 of 14 files at r9.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[classes/ObjectModel.php, line 520 at r9](https://beta.reviewable.io/reviews/prestashop/prestashop/8246#-L9oByOB2eWRe-p8nfUc:-L9oByOB2eWRe-p8nfUd:bb4kkmg) ([raw file](https://github.com/prestashop/prestashop/blob/9b7e14835bab4f97b82230a5cc2576fb7b1c7e48/classes/ObjectModel.php#L520)):*
> ```PHP
>         if (Shop::isTableAssociated($this->def['table'])) {
>             $id_shop_list = Shop::getContextListShopID();
>             if (count($this->id_shop_list) > 0) {
> ```

I'm not a fan of the implicit boolean conversion here, but it's not a blocker

---

*[tests/Unit/Adapter/Module/Configuration/ModuleSelfConfiguratorTest.php, line 276 at r9](https://beta.reviewable.io/reviews/prestashop/prestashop/8246#-L9oCdCN60Sk3IJAxlZw:-L9oCdCN60Sk3IJAxlZx:br67ab1) ([raw file](https://github.com/prestashop/prestashop/blob/9b7e14835bab4f97b82230a5cc2576fb7b1c7e48/tests/Unit/Adapter/Module/Configuration/ModuleSelfConfiguratorTest.php#L276)):*
> ```PHP
>     }
> 
>     public function delete($key)
> ```

This is a backwards-compatibility break that should at least be noted on the PR description

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/prestashop/prestashop/8246#-:-L9oCkOU4rDZq8bpoV2y:bnfp4nl)*

**Review by eternoendless — DISMISSED** [review]

**Comment by Quetzacoalt91** [comment]
*[tests/Unit/Adapter/Module/Configuration/ModuleSelfConfiguratorTest.php, line 276 at r9](https://beta.reviewable.io/reviews/prestashop/prestashop/8246#-L9oCdCN60Sk3IJAxlZw:-L9oFCjH5ObvR6ikd22E:b-ocltxi) ([raw file](https://github.com/prestashop/prestashop/blob/9b7e14835bab4f97b82230a5cc2576fb7b1c7e48/tests/Unit/Adapter/Module/Configuration/ModuleSelfConfiguratorTest.php#L276)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>

This is a backwards-compatibility break that should at least be noted on the PR description
</blockquote></details>

This is actually done in a PR already merged: https://github.com/PrestaShop/PrestaShop/pull/8705/files#diff-de0c0dfba5e525ab7feb53499a33ed01R175

:)

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/prestashop/prestashop/8246)*

**Review by mickaelandrieu — DISMISSED** [review]

**Comment by eternoendless** [comment]
@Quetzacoalt91 there are some rebase conflicts

**Comment by Quetzacoalt91** [comment]
Done

**Comment by eternoendless** [comment]
Reviewed 12 of 13 files at r10.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/prestashop/prestashop/8246)*

**Review by eternoendless — DISMISSED** [review]

**Review by LittleBigDev — DISMISSED** [review]

**Comment by eternoendless** [comment]
Reviewed 3 of 3 files at r11.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/prestashop/prestashop/8246)*

**Review by eternoendless — APPROVED** [review]

**Comment by eternoendless** [comment]
Thank you @Quetzacoalt91

**Pull request merged**